lib/document.js

```diff
@@ -3877,33 +3844,47 @@ Document.prototype.$toObject = function(options, json) {
   options.json = json;
   options.minimize = _minimize;
 
-  cloneOptions._parentOptions = options;
+  options._parentOptions = options;
 
-  cloneOptions._skipSingleNestedGetters = false;
+  options._skipSingleNestedGetters = false;
   // remember the root transform function
   // to save it from being overwritten by sub-transform functions
-  const originalTransform = options.transform;
+  // const originalTransform = options.transform;
 
-  let ret = clone(this._doc, cloneOptions) || {};
+  let ret = clone(this._doc, options) || {};
 
-  cloneOptions._skipSingleNestedGetters = true;
-  if (options.getters) {
-    applyGetters(this, ret, cloneOptions);
+  options._skipSingleNestedGetters = true;
+  const getters = options._calledWithOptions.getters
+    ?? options.getters
+    ?? defaultOptions.getters
+    ?? false;
+  if (getters) {
+    applyGetters(this, ret, options);
 
     if (options.minimize) {
       ret = minimize(ret) || {};
     }
   }
 
-  if (options.virtuals || (options.getters && options.virtuals !== false)) {
-    applyVirtuals(this, ret, cloneOptions, options);
+  const virtuals = options._calledWithOptions.virtuals
+    ?? defaultOptions.virtuals
+    ?? undefined;
```

Comment:
just to confirm, there is no `options.virtuals` that needs to be checked?

Reply:
No, because the order of precedence is 1) `toObject({ virtuals: true })`, 2) `schema.set('toObject', { virtuals: true })`, 3) `mongoose.set('toObject', { virtuals: true })`. (1) is from `_calledWithOptions`; (2) and (3) are rolled into `defaultOptions`.

`options` is primarily for calling `clone()`. `$toObject()` should primarily be concerned with `_calledWithOptions`